### PR TITLE
[BAHIR-305] Kudu Flink SQL Support DynamicSource/Sink&LookupFunction

### PR DESCRIPTION
**Pull request by collabH**
## ISSUE

ISSUE URL:[ISSUE](https://issues.apache.org/jira/projects/BAHIR/issues/BAHIR-305?filter=allissues)

## feature

* support flink dynamic source/sink factory
* upgrade kuduCatalogFactory to support dynamicFactory
* kudu catalog support hash bucket nums config
* Compatible with RowData and Row Type

### CODE REVIEW

**Comment by collabH** [comment]
@eskabetxe could you check?

**Comment by eskabetxe** [comment]
thanks @collabH for your contribution

@mbalassi could you check?

File: flink-connector-kudu/src/main/java/org/apache/flink/connectors/kudu/connector/convertor/RowResultRowDataConvertor.java [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+                case INT32:
+                case INT16:
+                case INT8:
+                    values.setField(pos, row.getInt(name));
+                    break;
+                case BOOL:
+                    values.setField(pos, row.getBoolean(name));
+                    break;
+                default:
+                    throw new IllegalArgumentException("columnName:" + name + ",type:" + type.getName() + "不支持!");
```
Comment (eskabetxe) [inline flink-connector-kudu/src/main/java/org/apache/flink/connectors/kudu/connector/convertor/RowResultRowDataConvertor.java:83]: can you translate to english the message?

File: flink-connector-kudu/src/main/java/org/apache/flink/connectors/kudu/connector/writer/RowDataUpsertOperationMapper.java [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+            case INTEGER:
+            case DATE:
+            case INTERVAL_YEAR_MONTH:
+                return input.getInt(i);
+            case TIME_WITHOUT_TIME_ZONE:
+                final int timePrecision = getPrecision(fieldType);
+                if (timePrecision < MIN_TIME_PRECISION || timePrecision > MAX_TIME_PRECISION) {
+                    throw new UnsupportedOperationException(
+                            String.format("The precision %s of TIME type is out of the range [%s, %s] supported by " +
+                                    "HBase connector", timePrecision, MIN_TIME_PRECISION, MAX_TIME_PRECISION));
```
Comment (eskabetxe) [inline flink-connector-kudu/src/main/java/org/apache/flink/connectors/kudu/connector/writer/RowDataUpsertOperationMapper.java:105]: HBase?

File: flink-connector-kudu/src/main/java/org/apache/flink/connectors/kudu/format/AbstractKuduInputFormat.java [hunk 3]
```diff
@@ -42,39 +43,42 @@
...
+        this(readerConfig, rowResultConvertor, tableInfo, new ArrayList<>(), tableProjections);
     }
 
-    public KuduRowInputFormat(KuduReaderConfig readerConfig, KuduTableInfo tableInfo, List<KuduFilterInfo> tableFilters, List<String> tableProjections) {
+    public AbstractKuduInputFormat(KuduReaderConfig readerConfig, RowResultConvertor<T> rowResultConvertor,
+                                   KuduTableInfo tableInfo, List<KuduFilterInfo> tableFilters,
+                                   List<String> tableProjections) {
 
         this.readerConfig = checkNotNull(readerConfig, "readerConfig could not be null");
+        this.rowResultConvertor = checkNotNull(rowResultConvertor, "readerConfig could not be null");
```
Comment (eskabetxe) [inline flink-connector-kudu/src/main/java/org/apache/flink/connectors/kudu/format/AbstractKuduInputFormat.java:81]: in message "readerConfig" should be rowResultConvertor

File: flink-connector-kudu/src/main/java/org/apache/flink/connectors/kudu/table/dynamic/KuduDynamicTableSourceSinkFactory.java [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+                    .intType()
+                    .defaultValue(KUDU_REPLICAS.defaultValue() * 2)
+                    .withDescription("kudu's hash partition bucket nums,defaultValue is 2 * replica nums");
+
+    public static final ConfigOption<String> KUDU_PRIMARY_KEY_COLS =
+            ConfigOptions
+                    .key("kudu.primary-key-columns")
+                    .stringType()
+                    .noDefaultValue()
+                    .withDescription("kudu's primary key,primary key must be ordered");
```
Comment (eskabetxe) [inline flink-connector-kudu/src/main/java/org/apache/flink/connectors/kudu/table/dynamic/KuduDynamicTableSourceSinkFactory.java:122]: space after comma ','

File: flink-connector-kudu/src/main/java/org/apache/flink/connectors/kudu/table/dynamic/KuduDynamicTableSourceSinkFactory.java [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+
+    /**
+     * hash partition bucket nums
+     */
+    public static final ConfigOption<Integer> KUDU_HASH_PARTITION_NUMS =
+            ConfigOptions
+                    .key("kudu.hash-partition-nums")
+                    .intType()
+                    .defaultValue(KUDU_REPLICAS.defaultValue() * 2)
+                    .withDescription("kudu's hash partition bucket nums,defaultValue is 2 * replica nums");
```
Comment (eskabetxe) [inline flink-connector-kudu/src/main/java/org/apache/flink/connectors/kudu/table/dynamic/KuduDynamicTableSourceSinkFactory.java:115]: space after comma ','

File: flink-connector-kudu/src/main/java/org/apache/flink/connectors/kudu/table/dynamic/KuduDynamicTableSourceSinkFactory.java [hunk 6]
```diff
@@ -0,0 +1,243 @@
...
+    public DynamicTableSource createDynamicTableSource(Context context) {
+        ReadableConfig config = getReadableConfig(context);
+        String masterAddresses = config.get(KUDU_MASTERS);
+
+        int scanRowSize = config.get(KUDU_SCAN_ROW_SIZE);
+        long kuduCacheMaxRows = config.get(KUDU_LOOKUP_CACHE_MAX_ROWS);
+        long kuduCacheTtl = config.get(KUDU_LOOKUP_CACHE_TTL);
+        int kuduMaxReties = config.get(KUDU_LOOKUP_MAX_RETRIES);
+
+        // 构造kudu lookup options
```
Comment (eskabetxe) [inline flink-connector-kudu/src/main/java/org/apache/flink/connectors/kudu/table/dynamic/KuduDynamicTableSourceSinkFactory.java:206]: can you translate?

**Review by eskabetxe — COMMENTED** [review]
@collabH sorry for late review..
I left some minor comments

File: flink-connector-kudu/src/main/java/org/apache/flink/connectors/kudu/connector/convertor/RowResultRowDataConvertor.java (hunk 1 above)
Comment (collabH) [reply flink-connector-kudu/src/main/java/org/apache/flink/connectors/kudu/connector/convertor/RowResultRowDataConvertor.java:83]: okay，I'll go fix it

File: flink-connector-kudu/src/main/java/org/apache/flink/connectors/kudu/connector/writer/RowDataUpsertOperationMapper.java (hunk 2 above)
Comment (collabH) [reply flink-connector-kudu/src/main/java/org/apache/flink/connectors/kudu/connector/writer/RowDataUpsertOperationMapper.java]: done

File: flink-connector-kudu/src/main/java/org/apache/flink/connectors/kudu/format/AbstractKuduInputFormat.java (hunk 3 above)
Comment (collabH) [reply flink-connector-kudu/src/main/java/org/apache/flink/connectors/kudu/format/AbstractKuduInputFormat.java]: done

File: flink-connector-kudu/src/main/java/org/apache/flink/connectors/kudu/table/dynamic/KuduDynamicTableSourceSinkFactory.java (hunk 6 above)
Comment (collabH) [reply flink-connector-kudu/src/main/java/org/apache/flink/connectors/kudu/table/dynamic/KuduDynamicTableSourceSinkFactory.java]: done

File: flink-connector-kudu/src/main/java/org/apache/flink/connectors/kudu/table/dynamic/KuduDynamicTableSourceSinkFactory.java (hunk 5 above)
Comment (collabH) [reply flink-connector-kudu/src/main/java/org/apache/flink/connectors/kudu/table/dynamic/KuduDynamicTableSourceSinkFactory.java]: done

File: flink-connector-kudu/src/main/java/org/apache/flink/connectors/kudu/table/dynamic/KuduDynamicTableSourceSinkFactory.java (hunk 4 above)
Comment (collabH) [reply flink-connector-kudu/src/main/java/org/apache/flink/connectors/kudu/table/dynamic/KuduDynamicTableSourceSinkFactory.java]: done

**Comment by collabH** [comment]
@eskabetxe hi,Can this pr be merged?

**Comment by eskabetxe** [comment]
@collabH merged.

**Pull request merged**